### DISCUSSION TITLE
Pimcore 11 | Codeception | missing the \Pimcore\Tests\Support\Util\Autoloader class / tests directory?

### DISCUSSION THREAD

**Discussion by jantomicky** [pimcore/pimcore#16511] [Q&A / Support] [answered]
Hello everyone,

I've just upgraded a Pimcore 6.9 project do Pimcore 11.1.4 and wanted to write a few API tests, so I followed the docs:
https://pimcore.com/docs/platform/Pimcore/Development_Tools_and_Details/Testing/Application_Testing#codeception

However, the IDE is warning me about a missing class (`\Pimcore\Tests\Support\Util\Autoloader`):

`Undefined class 'Autoloader'`

<img width="710" alt="Screenshot 2024-01-27 v 13 16 01" src="https://github.com/pimcore/pimcore/assets/41518085/fa59aba5-4d30-4493-a9d3-5fd08cdb7327">

I can see the class in the Pimcore repository: https://github.com/pimcore/pimcore/blob/11.x/tests/Support/Util/Autoloader.php

But my Pimcore Composer package is missing the whole `tests` directory:

<img width="468" alt="Screenshot 2024-01-27 v 13 33 11" src="https://github.com/pimcore/pimcore/assets/41518085/c910a251-c671-4883-807b-6097f2d201a2">

How so? Thanks in advance for any tips on this.

Edit 1:

```
"preferred-install": {
  "pimcore/pimcore": "source"
},
```

…does not seem to help anything.

Edit 2:

Related? https://github.com/pimcore/pimcore/discussions/16168

Edit 3:

I see now, the `tests/` directory is excluded in https://github.com/pimcore/pimcore/blob/11.x/.gitattributes. So what is the solution here? Why is it done this way, and should I create an issue regarding the docs?

**Comment by fashxp** [comment]
Hi,
in pimcore dist package, the tests folder is excluded. If you need it, make sure to tell composer to load from source instead, see for example https://github.com/pimcore/data-hub/blob/1.x/composer.json#L9

**Comment by jantomicky** [comment, marked as answer]
Thank you very much, @fashxp, you've helped me to find the issue. Composer was failing to install the source code for Pimcore because of this error:

> Failed to download pimcore/pimcore from source: git was not found in your PATH, skipping source download

I am using an Alpine based image, so there was no Git binary to handle the source installation. Now, after running:

```
rm -rf vendor/pimcore
composer clear-cache
composer update pimcore/pimcore --prefer source
```

in addition to using the configuration you've linked above, the `tests/` directory is no longer missing and I can use the `Pimcore\Tests\Support\Util\Autoloader` class no problem. Thanks again!
